src/core/gui/PageView.cpp

```diff
@@ -302,26 +306,43 @@ auto XojPageView::onButtonPressEvent(const PositionInputData& pos) -> bool {
     if (((h->getToolType() == TOOL_PEN || h->getToolType() == TOOL_HIGHLIGHTER) &&
          h->getDrawingType() != DRAWING_TYPE_SPLINE) ||
         (h->getToolType() == TOOL_ERASER && h->getEraserType() == ERASER_TYPE_WHITEOUT)) {
-        delete this->inputHandler;
-        this->inputHandler = nullptr;
-
-        if (h->getDrawingType() == DRAWING_TYPE_LINE) {
-            this->inputHandler = new RulerHandler(this->xournal, this, getPage());
-        } else if (h->getDrawingType() == DRAWING_TYPE_RECTANGLE) {
-            this->inputHandler = new RectangleHandler(this->xournal, this, getPage());
-        } else if (h->getDrawingType() == DRAWING_TYPE_ELLIPSE) {
-            this->inputHandler = new EllipseHandler(this->xournal, this, getPage());
-        } else if (h->getDrawingType() == DRAWING_TYPE_ARROW) {
-            this->inputHandler = new ArrowHandler(this->xournal, this, getPage(), false);
-        } else if (h->getDrawingType() == DRAWING_TYPE_DOUBLE_ARROW) {
-            this->inputHandler = new ArrowHandler(this->xournal, this, getPage(), true);
-        } else if (h->getDrawingType() == DRAWING_TYPE_COORDINATE_SYSTEM) {
-            this->inputHandler = new CoordinateSystemHandler(this->xournal, this, getPage());
-        } else {
-            this->inputHandler = new StrokeHandler(this->xournal, this, getPage());
+
+        if (this->inputHandler) {
+            this->deleteViewOf(this->inputHandler);
+            delete this->inputHandler;
+            this->inputHandler = nullptr;
+        }
+
+        BaseShapeHandler* shapeHandler = nullptr;
+        switch (h->getDrawingType()) {
+            case DRAWING_TYPE_LINE:
+                shapeHandler = new RulerHandler(this->xournal, this, getPage());
+                break;
+            case DRAWING_TYPE_RECTANGLE:
+                shapeHandler = new RectangleHandler(this->xournal, this, getPage());
+                break;
+            case DRAWING_TYPE_ELLIPSE:
+                shapeHandler = new EllipseHandler(this->xournal, this, getPage());
+                break;
+            case DRAWING_TYPE_ARROW:
+                shapeHandler = new ArrowHandler(this->xournal, this, getPage(), false);
+                break;
+            case DRAWING_TYPE_DOUBLE_ARROW:
+                shapeHandler = new ArrowHandler(this->xournal, this, getPage(), true);
+                break;
+            case DRAWING_TYPE_COORDINATE_SYSTEM:
+                shapeHandler = new CoordinateSystemHandler(this->xournal, this, getPage());
+                break;
+            default:
+                this->inputHandler = new StrokeHandler(this->xournal, this, getPage());
+        }
+        if (shapeHandler) {
+            this->inputHandler = shapeHandler;
+            this->overlayViews.emplace_back(std::make_unique<xoj::view::ShapeToolView>(shapeHandler, this));
```

Comment:
It really feels bad that we recreate all Handlers, when a button is pressed. This looks like a huge performance loss.
It would be better to either have a std::variant, containing all. Or have them all pre-allocated and only std::exchange them in.

Reply:
It's often safer to recreate, than to reset a structure, but I would just call something like this:

`*handler = Handler{};`
instead of `handler = new Handler{};`
or `new(handler) Handler{};`